From d6a15dddfed441faf0614ae7bde52e7a8103a624 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Wed, 17 Mar 2021 11:02:18 +0100
Subject: [PATCH 0487/1695] core: kernel: trace_ext: add printf function

printf is need to display clock tree without TEE CORE prefix.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 7152c539d79a11db6de79efeb961dc27b7e3d6cd core: kernel: trace_ext: add printf function]
Change-Id: I136ed27dc68c28a06dc38d8b4c9e5b16bee69c3d
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
[etienne: picked from rebase-3.19 development branch + remove 1 empty line]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/kernel/trace_ext.c | 52 +++++++++++++++++++++++++++++++++++++++--
 1 file changed, 50 insertions(+), 2 deletions(-)

diff --git a/core/kernel/trace_ext.c b/core/kernel/trace_ext.c
index 6ed3bf94c..6b8616324 100644
--- a/core/kernel/trace_ext.c
+++ b/core/kernel/trace_ext.c
@@ -2,13 +2,15 @@
 /*
  * Copyright (c) 2014, Linaro Limited
  */
-#include <stdbool.h>
-#include <trace.h>
 #include <console.h>
 #include <kernel/misc.h>
 #include <kernel/spinlock.h>
 #include <kernel/thread.h>
 #include <mm/core_mmu.h>
+#include <printk.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <trace.h>
 
 const char trace_ext_prefix[] = "TC";
 int trace_level __nex_data = TRACE_LEVEL;
@@ -61,3 +63,49 @@ int trace_ext_get_core_id(void)
 	else
 		return -1;
 }
+
+int printf(const char *fmt, ...)
+{
+	char to_format[MAX_PRINT_SIZE];
+	va_list ap;
+	int s;
+
+	if (trace_get_level() < TRACE_PRINTF_LEVEL)
+		return 0;
+
+	va_start(ap, fmt);
+	s = vsnprintk(to_format, sizeof(to_format), fmt, ap);
+	va_end(ap);
+
+	if (s < 0)
+		return s;
+
+	trace_ext_puts(to_format);
+
+	return s;
+}
+
+int putchar(int c)
+{
+	char str[2] = { (char)c, '\0' };
+
+	if (trace_get_level() >= TRACE_PRINTF_LEVEL)
+		trace_ext_puts(str);
+
+	/*
+	 * From the putchar() man page:
+	 * "fputc(), putc() and putchar() return the character written as an
+	 * unsigned char cast to an int or EOF on error."
+	 */
+	return (int)(unsigned char)c;
+}
+
+int puts(const char *str)
+{
+	if (trace_get_level() >= TRACE_PRINTF_LEVEL) {
+		trace_ext_puts(str);
+		trace_ext_puts("\n");
+	}
+
+	return 1;
+}
-- 
2.47.2

